### PR TITLE
mse loss added to experimental.

### PR DESCRIPTION
**Pull request by Suraj520**
Added mean squared loss to experimental.

### CODE REVIEW

**Comment by Suraj520** [comment]
@karalleyna - Awaiting your review. Kindly review and update :smile:

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py [hunk 1]
```diff
@@ -76,3 +76,81 @@ def test_binary_cross_entropy_with_logits(
         pos_weight=pos_weight[0],
         reduction=reduction,
     )
+
+#mse
+@handle_test(
+    fn_tree="mse_loss",
```
Comment (hirwa-nshuti) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:82]: You should specify the full function tree `functional.ivy.experimental` is missing

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey @Suraj520 

Thanks for the PR😊

The unit tests are failing firstlly due to the function tree issue and I am sure that the way you generated testing examples didn't properly handle the dimensions of both the true values and pred values for both the functions.

Also comment with `close #issue_number` in the first comment box of the PR to indicate subtasks that you are working on🙂

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py (hunk 1 above)
Comment (Suraj520) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:82]: Thanks for pointing out the glitch @hirwa-nshuti !

I've updated the fn_tree with full function tree i.e functional.ivy.experimental.mae_loss and functional.ivy.experimental.mse_loss

**Comment by Suraj520** [comment]
> Hey @Suraj520
> 
> Thanks for the PRblush
> 
> The unit tests are failing firstlly due to the function tree issue and I am sure that the way you generated testing examples didn't properly handle the dimensions of both the true values and pred values for both the functions.
> 
> Also comment with `close #issue_number` in the first comment box of the PR to indicate subtasks that you are working onslightly_smiling_face

@hirwa-nshuti - I've fixed all the quoted issues ! Kindly review.

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey would you update the branch with master branch to run the tests😊

**Comment by Suraj520** [comment]
> Hey would you update the branch with master branch to run the testsblush

Can you please check !

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py [hunk 2]
```diff
@@ -76,3 +76,91 @@ def test_binary_cross_entropy_with_logits(
...
+    test_flags,
+    backend_fw,
+    fn_name,
+    on_device,
+    ground_truth_backend,
+):
+    pred_dtype, pred = dtype_and_pred
+    true_dtype, true = dtype_and_true
+    # Add dimension handling
+    if len(pred.shape) > len(true.shape):
```
Comment (hirwa-nshuti) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:108]: As generated examples are lists you will need to unpack them

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
I tried to run your tests but they're failing due to improper unpacking of generated examples😊
Check the comment below

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py (hunk 2 above)
Comment (Suraj520) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:108]: @hirwa-nshuti - Thank you very much for your instant review :smile:  I am so sorry that I didn't run the pytest before committing. I shall do this and recommit with updated changes.

Request you to wait for some time :blush:  as I am a bit involved in some urgent tasks!

Thanks for your understanding.

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey @Suraj520 

Would you only add the changes related to the functions that you're implementing only because for now there failing tests for duplicacy in unit tests implementation

**Comment by Suraj520** [comment]
> Hey @Suraj520
> 
> Would you only add the changes related to the functions that you're implementing only because for now there failing tests for duplicacy in unit tests implementation


It seems that all the previous implementations in master repo in the path ivy/functional/ivy/experimental/losses.py & ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py have been migrated from experimental !

Now, The MSE and MAE loss and their respective losses have to be there ! I've fixed that ! Can you please review ?

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey,
There are still merge conflicts in this PR🙂
Would you look into it?

**Comment by Suraj520** [comment]
> Hey, There are still merge conflicts in this PRslightly_smiling_face Would you look into it?

So sorry for the delay, Could you please check now !

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
If you check the duplicate tests run on the CI is still failing this means that there are still duplicated functions you added to this PR, make sure that you're only adding mse loss.

**Comment by Suraj520** [comment]
> If you check the duplicate tests run on the CI is still failing this means that there are still duplicated functions you added to this PR, make sure that you're only adding mse loss.

It's weird @hirwa-nshuti :
Refer https://github.com/Suraj520/ivy/blob/loss_function/ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py and https://github.com/Suraj520/ivy/blob/loss_function/ivy/functional/ivy/experimental/losses.py

I've no duplicate functions. Just one for MSE and one for MAE as the thread started with !

I wonder why it's failing :disappointed:

**Comment by hirwa-nshuti** [comment]
> > If you check the duplicate tests run on the CI is still failing this means that there are still duplicated functions you added to this PR, make sure that you're only adding mse loss.
> 
> It's weird @hirwa-nshuti : Refer https://github.com/Suraj520/ivy/blob/loss_function/ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py and https://github.com/Suraj520/ivy/blob/loss_function/ivy/functional/ivy/experimental/losses.py
> 
> I've no duplicate functions. Just one for MSE and one for MAE as the thread started with !
> 
> I wonder why it's failing 😞

Go here at the changes on this PR https://github.com/unifyai/ivy/pull/14716/files you'll find the other losses tests added

**Comment by hirwa-nshuti** [comment]
> > If you check the duplicate tests run on the CI is still failing this means that there are still duplicated functions you added to this PR, make sure that you're only adding mse loss.
> 
> It's weird @hirwa-nshuti : Refer https://github.com/Suraj520/ivy/blob/loss_function/ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py and https://github.com/Suraj520/ivy/blob/loss_function/ivy/functional/ivy/experimental/losses.py
> 
> I've no duplicate functions. Just one for MSE and one for MAE as the thread started with !
> 
> I wonder why it's failing 😞

Go here at the changes on this PR https://github.com/unifyai/ivy/pull/14716/files you'll find the other losses tests added

**Comment by Suraj520** [comment]
> If you check the logs here you'll now see the cause of failures

Thanks for the assist ! It seems like it's stuck at inconsistent use of tabs error. This's a very disturbing persistent bug for python devs. I hope to fix it soon with patience to replace each tab with blankspaces in the code; or atleast get started with the line where it yields error. Thanks for your patience.

**Comment by Suraj520** [comment]
Thanks for your patience @hirwa-nshuti :smile: ! The duplicate-test-checks are sorted now !

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey @Suraj520 

Cool now there is no duplicate failure, would you add `ivy.Container` and `ivy.Array` instance methods of these functions as the tests are failing due to that😊

Thanks

**Comment by Suraj520** [comment]
> Hey @Suraj520
> 
> Cool now there is no duplicate failure, would you add `ivy.Container` and `ivy.Array` instance methods of these functions as the tests are failing due to thatblush
> 
> Thanks

@hirwa-nshuti : Added those but now the tests are failing at log_poisson_loss. Please check and suggest a fix !

```

ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py F [100%]

=================================== FAILURES ===================================
______ test_log_poisson_loss[cpu-ivy.functional.backends.jax-False-False] ______
ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:11: in test_log_poisson_loss
    fn_tree="functional.ivy.experimental.log_poisson_loss",
ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:49: in test_log_poisson_loss
    helpers.test_function(
ivy_tests/test_ivy/helpers/function_testing.py:307: in test_function
    target_fn = instance.__getattribute__(fn_name)
ivy/data_classes/array/array.py:389: in __getattribute__
    return super().__getattribute__(item)
E   AttributeError: 'Array' object has no attribute 'log_poisson_loss'
E   Falsifying example: test_log_poisson_loss(
E       backend_fw=<module 'ivy.functional.backends.jax' from '/ivy/ivy/functional/backends/jax/__init__.py'>,
E       on_device='cpu',
E       dtype_and_targets=(['bfloat16'], [array([0, 0, 0], dtype=bfloat16)]),
E       dtype_and_log_input=(['bfloat16'], [array([1, 1, 1], dtype=bfloat16)]),
E       compute_full_loss=True,
E       test_flags=FunctionTestFlags(
E           num_positional_args=0,
E           with_out=False,
E           instance_method=True,
E           test_gradients=None,
E           test_compile=None,
E           as_variable=[False],
E           native_arrays=[False],
E           container=[False],
E       ),
E       fn_name='log_poisson_loss',
E       ground_truth_backend='tensorflow',
E   )
```

**Comment by Suraj520** [comment]
Mistakenly closed it hence reopened

**Comment by hirwa-nshuti** [comment]
> > Hey @Suraj520
> > Cool now there is no duplicate failure, would you add `ivy.Container` and `ivy.Array` instance methods of these functions as the tests are failing due to thatblush
> > Thanks
> 
> @hirwa-nshuti : Added those but now the tests are failing at log_poisson_loss. Please check and suggest a fix !
> 
> ```
> 
> ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py F [100%]
> 
> =================================== FAILURES ===================================
> ______ test_log_poisson_loss[cpu-ivy.functional.backends.jax-False-False] ______
> ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:11: in test_log_poisson_loss
>     fn_tree="functional.ivy.experimental.log_poisson_loss",
> ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:49: in test_log_poisson_loss
>     helpers.test_function(
> ivy_tests/test_ivy/helpers/function_testing.py:307: in test_function
>     target_fn = instance.__getattribute__(fn_name)
> ivy/data_classes/array/array.py:389: in __getattribute__
>     return super().__getattribute__(item)
> E   AttributeError: 'Array' object has no attribute 'log_poisson_loss'
> E   Falsifying example: test_log_poisson_loss(
> E       backend_fw=<module 'ivy.functional.backends.jax' from '/ivy/ivy/functional/backends/jax/__init__.py'>,
> E       on_device='cpu',
> E       dtype_and_targets=(['bfloat16'], [array([0, 0, 0], dtype=bfloat16)]),
> E       dtype_and_log_input=(['bfloat16'], [array([1, 1, 1], dtype=bfloat16)]),
> E       compute_full_loss=True,
> E       test_flags=FunctionTestFlags(
> E           num_positional_args=0,
> E           with_out=False,
> E           instance_method=True,
> E           test_gradients=None,
> E           test_compile=None,
> E           as_variable=[False],
> E           native_arrays=[False],
> E           container=[False],
> E       ),
> E       fn_name='log_poisson_loss',
> E       ground_truth_backend='tensorflow',
> E   )
> ```

This function is not among the ones you added if other tests are passing do let me know so I can review your PR

**Comment by Suraj520** [comment]
> > > Hey @Suraj520
> > > Cool now there is no duplicate failure, would you add `ivy.Container` and `ivy.Array` instance methods of these functions as the tests are failing due to thatblush
> > > Thanks
> > 
> > 
> > @hirwa-nshuti : Added those but now the tests are failing at log_poisson_loss. Please check and suggest a fix !
> > ```
> > 
> > ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py F [100%]
> > 
> > =================================== FAILURES ===================================
> > ______ test_log_poisson_loss[cpu-ivy.functional.backends.jax-False-False] ______
> > ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:11: in test_log_poisson_loss
> >     fn_tree="functional.ivy.experimental.log_poisson_loss",
> > ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:49: in test_log_poisson_loss
> >     helpers.test_function(
> > ivy_tests/test_ivy/helpers/function_testing.py:307: in test_function
> >     target_fn = instance.__getattribute__(fn_name)
> > ivy/data_classes/array/array.py:389: in __getattribute__
> >     return super().__getattribute__(item)
> > E   AttributeError: 'Array' object has no attribute 'log_poisson_loss'
> > E   Falsifying example: test_log_poisson_loss(
> > E       backend_fw=<module 'ivy.functional.backends.jax' from '/ivy/ivy/functional/backends/jax/__init__.py'>,
> > E       on_device='cpu',
> > E       dtype_and_targets=(['bfloat16'], [array([0, 0, 0], dtype=bfloat16)]),
> > E       dtype_and_log_input=(['bfloat16'], [array([1, 1, 1], dtype=bfloat16)]),
> > E       compute_full_loss=True,
> > E       test_flags=FunctionTestFlags(
> > E           num_positional_args=0,
> > E           with_out=False,
> > E           instance_method=True,
> > E           test_gradients=None,
> > E           test_compile=None,
> > E           as_variable=[False],
> > E           native_arrays=[False],
> > E           container=[False],
> > E       ),
> > E       fn_name='log_poisson_loss',
> > E       ground_truth_backend='tensorflow',
> > E   )
> > ```
> 
> This function is not among the ones you added if other tests are passing do let me know so I can review your PR

Yes, It seems like they are passing @hirwa-nshuti ! Can you please review :smile:

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Comment by Suraj520** [comment]
> If you are working on an open task, please **edit** the PR description to link to the issue you've created.
> 
> For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).
> 
> Thank you hugs

@hirwa-nshuti : Is the message due to the cause that I mistakenly closed the issue and reopened it ? If yes, Kindly fix :smiley:

File: ivy/data_classes/array/experimental/losses.py [hunk 3]
```diff
@@ -1,6 +1,102 @@
...
 
 
 class _ArrayWithLossesExperimental(abc.ABC):
-    pass
+    def mse_loss(
+        self: ivy.Array,
+        pred: Union[ivy.Array, ivy.NativeArray],
+        /,
+        *,
+        axis: Optional[int] = None,
```
Comment (hirwa-nshuti) [inline ivy/data_classes/array/experimental/losses.py:13]: This will fail the test as axis argument isn't defined in the original function

Comment (Suraj520) [reply ivy/data_classes/array/experimental/losses.py:13]: Thank you very much @hirwa-nshuti for your valuable feedback, I look forward to correcting this as per your feedback. Thanks for your patience in this PR, throughout :smile:

Comment (Suraj520) [reply ivy/data_classes/array/experimental/losses.py:13]: Updated the original functions with axis argument, @hirwa-nshuti. Let's see.

File: ivy/data_classes/container/experimental/losses.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
+# global
+from typing import Optional, Union, List, Dict
+
+# local
+import ivy
+from ivy.data_classes.container.base import ContainerBase
+
+
+class _ContainerWithLosses(ContainerBase):
```
Comment (hirwa-nshuti) [inline ivy/data_classes/container/experimental/losses.py:9]: This should be renamed to indicated that it is an experimental instance
```suggestion
class _ContainerWithLossesExperimental(ContainerBase):

```

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
You will need to import the container class in the Container init file from the experimental Directory, from there run the tests to see the failures that you will get and fix them accordingly😊
Thanks

File: ivy/data_classes/container/experimental/losses.py (hunk 4 above)
Comment (Suraj520) [reply ivy/data_classes/container/experimental/losses.py]: Sure, @hirwa-nshuti : Thanks for the suggestion. I've resolved it

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py [hunk 5]
```diff
@@ -58,3 +58,114 @@ def test_log_poisson_loss(
...
+)
+def test_mse_loss(
+    dtype_and_true,
+    dtype_and_pred,
+    reduction,
+    test_flags,
+    backend_fw,
+    fn_name,
+    on_device,
+    ground_truth_backend,
```
Comment (hirwa-nshuti) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:85]: The ground truth fixture has been hidden we no longer need to specify it here

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey @Suraj520 

Just a small change and do let me know if tests are passing locally from your end🙂

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py (hunk 5 above)
Comment (Suraj520) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:85]: Sure @hirwa-nshuti. I'll update the function as per remark and run the unit tests. Shall update you when it's successful.

File: ivy/data_classes/container/experimental/losses.py [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+
+        >>> x = ivy.array([1, 2, 3])
+        >>> y = ivy.Container(a=ivy.array([2, 3, 4]), b=ivy.array([5, 6, 7]))
+        >>> x.mse_loss(y)
+        {
+            a: ivy.array(1.),
+            b: ivy.array(9.)
+        }
+        """
+        return self.cont_single_map_in_function(
```
Comment (hirwa-nshuti) [inline ivy/data_classes/container/experimental/losses.py:161]: This function should be calling the static method

File: ivy/data_classes/container/experimental/losses.py [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+
+        >>> x = ivy.array([1, 2, 3])
+        >>> y = ivy.Container(a=ivy.array([2, 3, 4]), b=ivy.array([5, 6, 7]))
+        >>> x.mae_loss(y)
+        {
+            a: ivy.array(1.),
+            b: ivy.array(2.)
+        }
+        """
+        return self.cont_single_map_in_function(
```
Comment (hirwa-nshuti) [inline ivy/data_classes/container/experimental/losses.py:323]: This also should be calling static method

**Review by hirwa-nshuti — COMMENTED** [review]
Hey @Suraj520

Would you check the comments below😊

File: ivy/data_classes/container/experimental/losses.py (hunk 7 above)
Comment (Suraj520) [reply ivy/data_classes/container/experimental/losses.py:323]: Thank you for your response and patience @hirwa-nshuti ! Sure, I'll fix it.

Comment (hirwa-nshuti) [reply ivy/data_classes/container/experimental/losses.py]: Yeah, it makes sense😊
All the best with preparations I'll try my best to see if I can successfully get them merged😊

Comment (Suraj520) [reply ivy/data_classes/container/experimental/losses.py]: Thanks a lot, @hirwa-nshuti !

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.

**Pull request closed without merging**